README.md
# utile.visuals
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/utile.visuals)](https://CRAN.R-project.org/package=utile.visuals)
[![Total Downloads](https://cranlogs.r-pkg.org/badges/grand-total/utile.visuals)](https://CRAN.R-project.org/package=utile.visuals)

## Overview
A small set of functions for making visuals with ggplot2. Includes minimalist themes with transparent backgrounds and a suite of tools for creating survival curves with risk tables.

## Functions
### > gg
- `ggrisktable()`: A simple wrapper function which calculates the numbers at risk for a survival model and a given set of time points then creates a ggplot2 table with them.

### > geom_
- `geom_stepconfint()`: Produces a step function confidence interval for survival curves. Essentially the `ggplot2::geom_step()` for confidence intervals which ggplot2 elects not to provide.

### > theme_
- `theme_basic()`: A ggplot2 theme for plots which removes most background elements.
- `theme_risk()`: A ggplot2 theme for risk tables which removes most background elements.

### > append_
- `append_table()`: Aligns axes and combines a ggplot2 plot and table into a single plot. Can handle legends.


NEWS.md
# utile.visuals 0.3.2
* Support for `linewidth` parameter introduced in `ggplot::` version 0.3.4.
* `theme_basic()` now has more consistent legend title formatting.
* Fix for a class check error in `ggrisktable()`

# utile.visuals 0.3.1
* Consolidation of themes into `theme_basic()` and `theme_risk()`. Line and text color can now be selected with the `base_color` parameter. Additionally, sizes of elements and text now scale correctly with the specified `base_size` parameter.
* Fix for `survfit0` class check issues in `ggrisktable()`.

# utile.visuals 0.3.0
* `connect_origin()` was removed as it redundantly replicated the function of `survival::survfit0()`. Thanks for the tip, Beth!

# utile.visuals 0.2.4
* `connect_origin()` now verifies whether a strata column is present.

# utile.visuals 0.2.3
* Added parameter to `ggrisktable()` to allow preemptive specification of text color.

# utile.visuals 0.2.2
* Fixed inconsistency within `theme_white()`.

# utile.visuals 0.2.1
* `append_table()`: Legend extraction can now be toggled with the new `extract.legend` option. Extraction may not be a desired behavior if the legend is already embedded within plot area.
* `ggrisktable()`:
  - New option `strata.order` added to allow reordering of strata in final table.
  - Now includes hard stops for invalid option data.
* Added `connect_origin` for connecting KM curves to a plot origin.

# utile.visuals 0.2.0
* Added `ggrisktable()` for creating a ggplot2 risk table from a survival::survfit() model.
* Added `theme_risk_black()` and `theme_risk_white()` for risk table formatting.
* Added `append_table()` for adding a ggplot2 risk table to a ggplot2 Kaplan Meier curve.
* `theme_black()`|`theme_white()`: Updated axis title formating.
* Updated documentation.

# utile.visuals 0.1.1
* First public release


R/geoms.R
#' @title Step function confidence intervals for ggplot2
#' @description Produces a step function confidence interval for survival curves. Essentially
#' the geom_step() for confidence intervals which ggplot2 does not provide.
#' @param mapping Aesthetic mappings with aes() function. Like geom_ribbon(), you must provide
#' columns for x, ymin (lower limit), ymax (upper limit).
#' @param data The data to be displayed in this layer. Can inherit from ggplot parent.
#' @param stat The statistical transformation to use on the data for this layer, as a string.
#' Defaults to 'identity'.
#' @param position Position adjustment, either as a string, or the result of a call to a
#' position adjustment function.
#' @param na.rm If FALSE, the default, missing values are removed with a warning. If TRUE,
#' missing values are silently removed.
#' @param ... Optional. Any other ggplot geom_ribbon() arguments.
#' @note Adapted from the survminer package <https://github.com/kassambara/survminer>.
#' @examples
#' library(survival)
#' library(broom)
#' library(ggplot2)
#'
#' fit <- survfit(Surv(time, status) ~ trt, data = diabetic)
#' fit <- survfit0(fit) # connect origin
#'
#' ggplot(
#'   data = tidy(fit),
#'   mapping = aes(x = time, y = estimate)
#' ) +
#'   geom_step(aes(color = strata)) +
#'   geom_stepconfint(aes(ymin = conf.low, ymax = conf.high, fill = strata), alpha = 0.3) +
#'   coord_cartesian(c(0, 50)) +
#'   scale_x_continuous(expand = c(0.02,0)) +
#'   labs(x = 'Time', y = 'Freedom From Event') +
#'   scale_color_manual(
#'     values = c('#d83641', '#1A45A7'),
#'     name = 'Treatment',
#'     labels = c('None', 'Laser'),
#'     aesthetics = c('colour', 'fill')) +
#'   theme_basic()
#' @export
geom_stepconfint <- function (
  mapping = NULL, data = NULL, stat = "identity",
  position = "identity", na.rm = FALSE, ...
) {
  ggplot2::layer(
    mapping = mapping,
    data = data,
    stat = stat,
    geom = ggplot2::ggproto(
      `_class` = 'GeomConfint',
      `_inherit` = ggplot2::GeomRibbon,
      required_aes = c("x", "ymin", "ymax"),
      draw_group = function (self, data, panel_scales, coord, na.rm = FALSE) {
        if (na.rm) data <- data[stats::complete.cases(self$required_aes), ]
        data <- data[order(data$group, data$x), ]
        data <- self$stairstep_confint(data)
        ggplot2::GeomRibbon$draw_group(data, panel_scales, coord, na.rm = FALSE)
      },
      stairstep_confint = function (data) {
        data <- as.data.frame(data)[order(data$x), ]
        n <- nrow(data)
        ys <- rep(1:n, each = 2)[-2 * n]
        xs <- c(1, rep(2:n, each = 2))
        data.frame(
          x = data$x[xs],
          ymin = data$ymin[ys],
          ymax = data$ymax[ys],
          data[xs, setdiff(names(data), c("x", "ymin", "ymax"))]
        )
      }
    ),
    position = position,
    params = list(na.rm = na.rm, ...)
  )
}


R/append.R
#' @title Append a ggplot2 table to the bottom of a ggplot2 plot
#' @description Aligns axes and combines a ggplot2 plot and table into a single plot.
#' Can handle legends.
#' @param plot Required. ggplot2::ggplot() object. If a legend is present, it will be
#' extracted.
#' @param table Required. ggplot2::ggplot object. If a legend is present, it will
#' be removed and ignored.
#' @param plot.height Optional. Numeric. Height of plot relative to table. Defaults
#' to 1.
#' @param table.height Optional. Numeric. Height of table relative to plot. Defaults
#' to 0.1.
#' @param plot.width Optional. Numeric. Width of plot relative to legend. Ignored
#' if no legend present in plot. Defaults to 1.
#' @param extract.legend Optional. Logical. Indicates whether to extract the legend
#' from the plot and reinsert it adjacent to the final combined plot. May be undesired
#' if legend already embedded within the plot area. Defaults to TRUE.
#' @param legend.width Optional. Numeric. Width of legend relative to plot. Ignored
#' if no legend present in plot or 'extract.legend'=FALSE. Defaults 0.2.
#' @param legend.offset Optional. Numeric. Vertical offset of legend. Used to raise
#' or lower. Ignored if no legend present in plot or 'extract.legend'=FALSE. Defaults
#' to -15.
#' @return A ggplot2 tableGrob object. Use grid::grid.draw() to open in RStudio viewer.
#' Works with ggplot2::ggsave() out of the box.
#' @note To ensure proper alignment, double check that both plots use the same scale
#' and breaks!
#' @examples
#' library(survival)
#' library(ggplot2)
#' library(broom) # tidy() model data
#' library(grid) # grid.draw() finished plot
#'
#' # Data with group names specified
#' data_diabetic <- diabetic
#' data_diabetic$trt <- as.factor(data_diabetic$trt)
#' levels(data_diabetic$trt) <- c('None', 'Laser')
#'
#' # Survival Model
#' fit <- survfit(Surv(time, status) ~ trt, data = data_diabetic)
#'
#' # Kaplan Meier (KM) Plot
#' plot_km <- ggplot(
#'  data = tidy(fit),
#'  mapping = aes(x = time, y = estimate)
#' ) +
#'   geom_step(aes(color = strata)) +
#'   geom_stepconfint(aes(ymin = conf.low, ymax = conf.high, fill = strata), alpha = 0.3) +
#'   coord_cartesian(c(0, 50)) + # Note scale set here!
#'   scale_x_continuous(expand = c(0.02,0)) +
#'   labs(x = 'Time', y = 'Freedom From Event') +
#'   scale_color_manual(
#'     values = c('#d83641', '#1A45A7'),
#'     name = 'Treatment',
#'     labels = c('Laser', 'None'),
#'     aesthetics = c('colour', 'fill')) +
#'   theme_basic()
#'
#' # Risk Table
#' tbl_risk <- ggrisktable(fit, c(0, 10, 20, 30, 40, 50)) +
#'   coord_cartesian(c(0, 50)) +
#'   scale_x_continuous(expand = c(0.02,0)) +
#'   theme_risk()
#'
#' # Combine KM plot and risk table
#' plot_cmbd <- append_table(
#'   plot = plot_km,
#'   table = tbl_risk
#' )
#'
#' # Draw in RStudio viewer
#' grid.newpage()
#' grid.draw(plot_cmbd)
#' @export
append_table <- function(
  plot = NULL, table = NULL, plot.height = 1,
  table.height = 0.1, plot.width = 1, extract.legend = TRUE,
  legend.width = 0.2, legend.offset = -15
) {

  # Hard stops
  if (!ggplot2::is.ggplot(plot) | !ggplot2::is.ggplot(table))
    stop('Missing valid ggplot object. Check [\'plot\', \'table\'].')
  if (!is.numeric(plot.height) | !is.numeric(table.height))
    stop('Both plot and table heights must be specified. Check \'plot.height\', \'table.height\'.')

  # Convert plots to grobs
  if (extract.legend) plot <- plot +
      ggplot2::theme(
        legend.margin = ggplot2::margin(legend.offset, 0, 0, 0, unit = 'mm'),
        legend.justification = NULL,
        legend.position = NULL
      )
  grob_plot <- ggplot2::ggplotGrob(plot)
  grob_tbl <- ggplot2::ggplotGrob(table + ggplot2::theme(legend.position = 'none')) # drop table legend

  # Extract legend, if present
  if (extract.legend) {
    legend_row <- which(purrr::map_chr(grob_plot$grobs, ~ .x$name) == 'guide-box')
    if (length(legend_row) > 0) {

      # Hard stop
      if (!is.numeric(plot.width) | !is.numeric(legend.width))
        stop('Widths for both plot & legend required. Check [\'legend.width\', \'plot.width\'].')

      grob_legend <- grob_plot$grobs[[legend_row]]
      grob_plot <- ggplot2::ggplotGrob(plot + ggplot2::theme(legend.position = 'none'))
    }
  }

  # Combine plot grobs
  grob_combined <- rbind(
    grob_plot,
    grob_tbl,
    size = 'first'
  )
  panels <- grob_combined$layout$t[grep("panel", grob_combined$layout$name)]
  grob_combined$heights[panels[1]] <- ggplot2::unit(plot.height, "null") # Set plot height
  grob_combined$heights[panels[2]] <- ggplot2::unit(table.height, "null") # Set table height

  # Return new arrangement
  if (extract.legend & exists('grob_legend'))
    gridExtra::arrangeGrob(
      grob_combined,
      grob_legend,
      ncol = 2,
      widths = c(plot.width, legend.width),
      clip = TRUE
    )
  else grob_combined
}


R/themes.R
#' @importFrom ggplot2 %+replace%

#' @title Minimalist theme for ggplot2
#' @description A ggplot2 theme which removes most background
#' elements.
#' @param base_size Optional. Numeric. Base font size. Defaults to 14.
#' @param base_family Optional. Numeric. Base font family. Defaults to ?.
#' @param base_color Optional. Character. Base color for lines and text. Defaults to black.
#' @param base_line_size Optional. Numeric. Base line element size.
#' Defaults to base_size/22.
#' @param base_rect_size Optional. Numeric. Base rectangle element size.
#' Defaults to base_size/22.
#' @note Recommend exporting as PNG or TIFF to preserve
#' background transparency.
#' @export
theme_basic <- function(
  base_size = 12,
  base_family = "",
  base_color = "black",
  base_line_size = base_size/14,
  base_rect_size = base_size/14
) {
  ggplot2::theme_bw(
    base_size = base_size,
    base_family = base_family,
    base_line_size = base_line_size,
    base_rect_size = base_rect_size
  ) %+replace%
    ggplot2::theme(

      # Defaults
      text = ggplot2::element_text(
        family = base_family,
        color = base_color,
        size = base_size,
        lineheight = 0.9,
        hjust = 0.5,
        vjust = 0.5,
        angle = 0,
        margin = ggplot2::margin(),
        debug = FALSE),
      line = ggplot2::element_line(
        color = base_color,
        linewidth = base_line_size,
        linetype = 1,
        lineend = "square"),
      rect = ggplot2::element_rect(
        fill = "transparent",
        colour = base_color,
        linewidth = base_rect_size,
        linetype = 1),

      # Axis
      axis.line = ggplot2::element_line(
        colour = base_color, linewidth = base_line_size),
      axis.line.x = NULL,
      axis.line.y = NULL,
      axis.ticks = ggplot2::element_line(color = base_color),
      axis.ticks.length = ggplot2::unit(base_size / 2.5, "pt"),
      axis.ticks.length.x = NULL,
      axis.ticks.length.x.top = NULL,
      axis.ticks.length.x.bottom = NULL,
      axis.ticks.length.y = NULL,
      axis.ticks.length.y.left = NULL,
      axis.ticks.length.y.right = NULL,
      axis.title = ggplot2::element_text(color = base_color),
      axis.title.x = ggplot2::element_text(
        angle = 0, margin = ggplot2::margin(t = base_size * 0.6), vjust = 1),
      axis.title.x.top = ggplot2::element_text(
        margin = ggplot2::margin(b = base_size * 0.6), vjust = 0),
      axis.title.y = ggplot2::element_text(
        angle = 90, margin = ggplot2::margin(r = base_size * 0.6), vjust = 1),
      axis.title.y.right = ggplot2::element_text(
        angle = -90, margin = ggplot2::margin(l = base_size * 0.6), vjust = 0),
      axis.text = ggplot2::element_text(
        size = ggplot2::rel(0.95), color = base_color),
      axis.text.x = ggplot2::element_text(
        margin = ggplot2::margin(t = 0.8 * base_size / 4), vjust = 0.5),
      axis.text.x.top = ggplot2::element_text(
        margin = ggplot2::margin(b = 0.8 * base_size / 4),vjust = 0),
      axis.text.y = ggplot2::element_text(
        margin = ggplot2::margin(r = 0.8 * base_size / 4), hjust = 0.5),
      axis.text.y.right = ggplot2::element_text(
        margin = ggplot2::margin(l = 0.8 * base_size / 4), hjust = 0),

      # Legend
      legend.title = ggplot2::element_text(
        color = base_color, margin = ggplot2::margin(b = base_size * 0.4)),
      legend.title.align = 0,
      legend.text = ggplot2::element_text(
        color = base_color, size = ggplot2::rel(0.8)),
      legend.background = ggplot2::element_blank(),
      legend.key = ggplot2::element_blank(),
      legend.key.size = ggplot2::unit(1.2, "lines"),
      legend.key.width = ggplot2::unit(base_size * 1.8, "pt"),
      legend.spacing = ggplot2::unit(base_size, "pt"),
      legend.spacing.x = NULL,
      legend.spacing.y = NULL,
      legend.margin = ggplot2::margin(
        base_size/2, base_size/2, base_size/2, base_size/2),
      legend.position = "right",
      legend.justification = "center",
      legend.text.align = NULL,
      legend.direction = NULL,
      legend.box = NULL,
      legend.box.margin = ggplot2::margin(0, 0, 0, 0, "cm"),
      legend.box.background = ggplot2::element_blank(),
      legend.box.spacing = ggplot2::unit(base_size, "pt"),

      # Strip
      strip.background = ggplot2::element_rect(
        fill = "transparent", colour = base_color, linewidth = ggplot2::rel(2)),
      strip.text = ggplot2::element_text(
        color = base_color,
        size = ggplot2::rel(0.8),
        margin = ggplot2::margin(base_size/2.5, base_size/2.5, base_size/2.5, base_size/2.5)),
      strip.text.x = ggplot2::element_text(
        margin = ggplot2::margin(b = base_size / 3)),
      strip.text.y = ggplot2::element_text(
        angle = -90, margin = ggplot2::margin(l = base_size / 3)),
      strip.text.y.left = ggplot2::element_text(angle = 90),
      strip.placement = "inside",
      strip.placement.x = NULL,
      strip.placement.y = NULL,
      strip.switch.pad.grid = ggplot2::unit(base_size / 4, "pt"),
      strip.switch.pad.wrap = ggplot2::unit(base_size / 4, "pt"),

      # Panel
      panel.border = ggplot2::element_blank(),
      panel.grid = ggplot2::element_blank(),
      panel.grid.minor = ggplot2::element_blank(),
      panel.grid.major = ggplot2::element_blank(),
      panel.background = ggplot2::element_blank(),
      panel.spacing = ggplot2::unit(base_size / 2, "pt"),
      panel.spacing.x = NULL,
      panel.spacing.y = NULL,
      panel.ontop = FALSE,

      # Plot
      plot.background = ggplot2::element_rect(fill = "transparent", colour = NA),
      plot.title = ggplot2::element_text(
        size = ggplot2::rel(1.2), hjust = 0.5, vjust = 1,
        margin = ggplot2::margin(b = base_size)
      ),
      plot.title.position = "panel",
      plot.margin = ggplot2::margin(base_size/2, base_size/2, base_size/2, base_size/2),

      complete = TRUE
    )
}


#' @title Minimalist risk table theme for ggplot2
#' @description A ggplot2 theme which removes most background
#' elements.
#' @param base_size Optional. Numeric. Base font size. Defaults to 10.
#' @param base_family Optional. Numeric. Base font family. Defaults to ?.
#' @param base_color Optional. Character. Base color for lines and text. Defaults to black.
#' @param base_line_size Optional. Numeric. Base line element size.
#' Defaults to base_size/22.
#' @param base_rect_size Optional. Numeric. Base rectangle element size.
#' Defaults to base_size/22.
#' @note Recommend exporting as PNG or TIFF to preserve
#' background transparency.
#' @export
theme_risk <- function(
  base_size = 12,
  base_family = "",
  base_color = "black",
  base_line_size = base_size/14,
  base_rect_size = base_size/14
) {
  ggplot2::theme_bw(
    base_size = base_size,
    base_family = base_family,
    base_line_size = base_line_size,
    base_rect_size = base_rect_size
  ) %+replace%
    ggplot2::theme(

      # Defaults
      text = ggplot2::element_text(
        family = base_family,
        color = base_color,
        size = base_size,
        lineheight = 0.9,
        hjust = 0.5,
        vjust = 0.5,
        angle = 0,
        margin = ggplot2::margin(),
        debug = FALSE),
      line = ggplot2::element_line(
        color = base_color,
        linewidth = base_line_size,
        linetype = 1,
        lineend = "square"),
      rect = ggplot2::element_rect(
        fill = "transparent",
        colour = base_color,
        linewidth = base_rect_size,
        linetype = 1),

      # Axis
      axis.line = ggplot2::element_blank(),
      axis.line.x = NULL,
      axis.line.y = NULL,
      axis.ticks = ggplot2::element_blank(),
      axis.title = ggplot2::element_blank(),
      axis.title.x = NULL,
      axis.title.x.top = NULL,
      axis.title.y = NULL,
      axis.title.y.right = NULL,
      axis.text = ggplot2::element_text(
        size = ggplot2::rel(0.95), color = base_color),
      axis.text.x = ggplot2::element_blank(),
      axis.text.x.top = ggplot2::element_blank(),
      axis.text.y = ggplot2::element_text(
        margin = ggplot2::margin(r = 0.8 * base_size / 4), hjust = 0.5),
      axis.text.y.right = ggplot2::element_text(
        margin = ggplot2::margin(l = 0.8 * base_size / 4), hjust = 0),

      # Legend
      legend.title = ggplot2::element_blank(),
      legend.title.align = 0,
      legend.text = ggplot2::element_text(
        color = base_color, size = ggplot2::rel(0.8)),
      legend.background = ggplot2::element_blank(),
      legend.key = ggplot2::element_blank(),
      legend.key.size = ggplot2::unit(1.2, "lines"),
      legend.key.width = ggplot2::unit(base_size * 1.8, "pt"),
      legend.spacing = ggplot2::unit(base_size, "pt"),
      legend.spacing.x = NULL,
      legend.spacing.y = NULL,
      legend.margin = ggplot2::margin(
        base_size/2, base_size/2, base_size/2, base_size/2),
      legend.position = "right",
      legend.justification = "center",
      legend.text.align = NULL,
      legend.direction = NULL,
      legend.box = NULL,
      legend.box.margin = ggplot2::margin(0, 0, 0, 0, "cm"),
      legend.box.background = ggplot2::element_blank(),
      legend.box.spacing = ggplot2::unit(base_size, "pt"),

      # Strip
      strip.background = ggplot2::element_rect(
        fill = "transparent", colour = base_color, linewidth = ggplot2::rel(2)),
      strip.text = ggplot2::element_text(
        color = base_color,
        size = ggplot2::rel(0.8),
        margin = ggplot2::margin(base_size/2.5, base_size/2.5, base_size/2.5, base_size/2.5)),
      strip.text.x = ggplot2::element_text(
        margin = ggplot2::margin(b = base_size / 3)),
      strip.text.y = ggplot2::element_text(
        angle = -90, margin = ggplot2::margin(l = base_size / 3)),
      strip.text.y.left = ggplot2::element_text(angle = 90),
      strip.placement = "inside",
      strip.placement.x = NULL,
      strip.placement.y = NULL,
      strip.switch.pad.grid = ggplot2::unit(base_size / 4, "pt"),
      strip.switch.pad.wrap = ggplot2::unit(base_size / 4, "pt"),

      # Panel
      panel.border = ggplot2::element_blank(),
      panel.grid = ggplot2::element_blank(),
      panel.grid.minor = ggplot2::element_blank(),
      panel.grid.major = ggplot2::element_blank(),
      panel.background = ggplot2::element_blank(),
      panel.spacing = ggplot2::unit(base_size / 2, "pt"),
      panel.spacing.x = NULL,
      panel.spacing.y = NULL,
      panel.ontop = FALSE,

      # Plot
      plot.background = ggplot2::element_rect(fill = "transparent", colour = NA),
      plot.title = ggplot2::element_text(
        size = ggplot2::rel(1.2), hjust = 0.5, vjust = 1,
        margin = ggplot2::margin(b = base_size)
      ),
      plot.title.position = "panel",
      plot.margin = ggplot2::margin(base_size/1.5, base_size/2, base_size/2, base_size/2),

      complete = TRUE
    )
}


R/gg.R
utils::globalVariables(c('time', 'strata', 'n.risk'))

#' @title Create a ggplot2 table showing the number at risk
#' @description A simple wrapper function which calculates the
#' numbers at risk for a survival model and a given set of time points then
#' creates a ggplot2 table with them.
#' @param fit Required. survival::survfit() object.
#' @param times Required. Numeric. One or more time points to calculate
#' the number at risk for.
#' @param text.color Optional. Character. Color of text within table. Defaults
#' to 'black'.
#' @param strata.order Optional. Character. Ordered names of strata factor
#' levels.
#' @return An unformatted ggplot2 table showing the number at risk.
#' @examples
#' library(survival)
#'
#' fit <- survfit(Surv(time, status) ~ trt, data = diabetic)
#'
#' ggrisktable(
#'    fit = fit,
#'    times = c(0, 10, 20, 30, 40, 50),
#'    strata.order = c('0', '1')
#' ) + theme_risk()
#' @export
ggrisktable <- function (fit = NULL, times = NULL, text.color = 'black', strata.order = NULL) {

  # Hard stops
  if (is.null(fit) | !(inherits(fit, 'survfit'))) stop('No valid fit object provided. [Check: \'fit\']')
  if (is.null(times) | !is.numeric(times)) stop('No valid time points provided. [Check: \'times\']')
  if (!is.null(strata.order) & !is.character(strata.order)) stop('Invalid strata order data provided. [Check: \'strata.order\']')

  # Generate risk table and order
  risk_table <- .tabulate_at_risk(fit, times)

  # Reorder strata
  if (is.character(strata.order))
    risk_table$strata <- factor(
      risk_table$strata,
      levels = unique(c(
        rev(strata.order[strata.order %in% levels(risk_table$strata)]),
        levels(risk_table$strata)
      ))
    )

  # Return plotted table
  ggplot2::ggplot(
    risk_table,
    ggplot2::aes(x = time, y = strata, label = n.risk)
  ) + ggplot2::geom_text(color = text.color)
}
